### ISSUE TITLE
parse( ) throws OverflowError on negative binary values

### ISSUE THREAD

**Issue by ethomag** [JuliaLang/julia#37899]
```
julia> parse(Int16, bitstring(Int16(-1)), base=2)
ERROR: OverflowError: overflow parsing "1111111111111111"
Stacktrace:
 [1] tryparse_internal(::Type{Int16}, ::String, ::Int64, ::Int64, ::Int64, ::Bool) at ./parse.jl:160
 [2] parse(::Type{Int16}, ::String; base::Int64) at ./parse.jl:238
 [3] top-level scope at REPL[30]:1

julia> versioninfo()
Julia Version 1.5.2
Commit 539f3ce943 (2020-09-23 23:17 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) Silver 4114 CPU @ 2.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake-avx512)
```

**Comment by ethomag** [comment]
Aha, I suppose I should use a minus sign for negative values.  So, maybe this is not a bug after all.  For my use at hand, the parse function is less useful as defined.  But, I can understand that this makes sense in general:
```
julia> parse(Int16, "32768", base=10)
ERROR: OverflowError: overflow parsing "32768"
Stacktrace:
 [1] tryparse_internal(::Type{Int16}, ::String, ::Int64, ::Int64, ::Int64, ::Bool) at ./parse.jl:160
 [2] parse(::Type{Int16}, ::String; base::Int64) at ./parse.jl:238
 [3] top-level scope at REPL[47]:1
julia> parse(Int16, "-32768", base=10)
-32768
```

**Comment by StefanKarpinski** [comment]
The `bitstring` function does not print a representation that can necessarily be parsed by `parse`, it prints the sequence of bits that represent the value. It happens that for positive integers, that is the binary string representation of the value, but as you've discovered, for negative values, that's not the case. It's even further from the case if you use `bitstring` on floating-point values:
```jl
julia> x = 1.5
1.5

julia> bitstring(x)
"0011111111111000000000000000000000000000000000000000000000000000"
```
TLDR: `parse` is not the inverse of `bitstring`.

**Comment by ethomag** [comment]
Got it! Thanks for a clear explanation.  A follow-up question: is there an inverse of `bitstring`, if I want to process binary two's complement representations of integers ?

**Comment by ethomag** [comment]
Sorry for bugging you, without thinking first.  I can always do:
```
julia> reinterpret(Int16, parse(UInt16, bitstring(Int16(-1)), base=2))
-1
```

**Comment by oscardssmith** [comment]
Note that while this works, if you care at all about performance, the best way to work with bits of an int is to use shifts and masks. Parsing and converting to string is relatively expensive.

**Comment by StefanKarpinski** [comment]
Yes, that would work. The ideal is not to work with string representations of bits in the first place, but if that's what you have for some reason then you can do the parse and reinterpret thing you just wrote. Even works for floating-point:
```jl
julia> reinterpret(Float32, parse(UInt32, bitstring(1.5f0), base=2))
1.5f0
```
